java-checks/src/main/java/org/sonar/java/checks/regex/RegexComplexityCheck.java

```diff
@@ -0,0 +1,299 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.regex;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.apache.commons.lang.StringUtils;
+import org.sonar.check.Rule;
+import org.sonar.check.RuleProperty;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.java.regex.RegexParseResult;
+import org.sonar.java.regex.ast.BackReferenceTree;
+import org.sonar.java.regex.ast.CharacterClassIntersectionTree;
+import org.sonar.java.regex.ast.CharacterClassTree;
+import org.sonar.java.regex.ast.DisjunctionTree;
+import org.sonar.java.regex.ast.FlagSet;
+import org.sonar.java.regex.ast.JavaCharacter;
+import org.sonar.java.regex.ast.LookAroundTree;
+import org.sonar.java.regex.ast.NonCapturingGroupTree;
+import org.sonar.java.regex.ast.RegexBaseVisitor;
+import org.sonar.java.regex.ast.RegexSyntaxElement;
+import org.sonar.java.regex.ast.RegexToken;
+import org.sonar.java.regex.ast.RepetitionTree;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.SyntaxTrivia;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S5843")
+public class RegexComplexityCheck extends AbstractRegexCheck {
+
+  private static final String MESSAGE = "Simplify this regular expression to reduce its complexity from %d to the %d allowed.";
+
+  private static final int DEFAULT_MAX = 20;
+
+  @RuleProperty(
+    key = "maxComplexity",
+    description = "The maximum authorized complexity.",
+    defaultValue = "" + DEFAULT_MAX)
+  private int max = DEFAULT_MAX;
+
+  private final List<RegexConstructionInfo> regexConstructions = new ArrayList<>();
+
+  private final Set<Integer> commentedLines = new HashSet<>();
```

Comment:
you need to clear the list when leaving the file. It's the same instance of the rule which is used for all the checks. If not cleared, you store all the comments encountered along the analysis. To do so, simply override `leaveFile()` method as follow:

```
  @Override
  public void leaveFile(JavaFileScannerContext context) {
    commentedLines.clear();
  }
```

Reply:
Note that it might be the reason why sanity fails.